File: Linux/spiderip.cpp
/*
 * Title:  spider spiderip.cpp (Linux)
 * Author: Shuichiro Endo
 */

#include "spider.hpp"
#include "spiderip.hpp"


namespace spider
{
    Spiderip::Spiderip(std::string spider_ipv4,
                       std::string spider_ipv6_global,
                       std::string spider_ipv6_unique_local,
                       std::string spider_ipv6_link_local)
    {
        this->spider_ipv4 = spider_ipv4;
        this->spider_ipv6_global = spider_ipv6_global;
        this->spider_ipv6_unique_local = spider_ipv6_unique_local;
        this->spider_ipv6_link_local = spider_ipv6_link_local;

        if(!spider_ipv6_link_local.empty())    // scope_id
        {
            if(spider_ipv6_link_local.find("%") == std::string::npos)
            {
                std::printf("[-] ipv6 link local address does not include a scope id\n");

                init_flag = false;
            }else
            {
                this->spider_ipv6_link_local_scope_id = strstr(this->spider_ipv6_link_local.c_str(), "%") + 1;
                if(this->spider_ipv6_link_local_scope_id.empty())
                {
                    std::printf("[-] ipv6 link local address does not include a scope id\n");

                    init_flag = false;
                }else
                {
                    memcpy(this->addr6_string_pointer,
                           this->spider_ipv6_link_local.c_str(),
                           this->spider_ipv6_link_local.size() - this->spider_ipv6_link_local_scope_id.size() - 1);
                    this->spider_ipv6_link_local = this->addr6_string_pointer;

                    init_flag = true;
                }
            }
        }else
        {
            init_flag = true;
        }
    }

    Spiderip::~Spiderip()
    {

    }

    void Spiderip::set_init_flag(bool init_flag)
    {
        this->init_flag = init_flag;
    }

    bool Spiderip::get_init_flag()
    {
        return init_flag;
    }

    void Spiderip::set_spider_ipv4(std::string spider_ipv4)
    {
        this->spider_ipv4 = spider_ipv4;
    }

    std::string Spiderip::get_spider_ipv4()
    {
        return spider_ipv4;
    }

    void Spiderip::set_spider_ipv6_global(std::string spider_ipv6_global)
    {
        this->spider_ipv6_global = spider_ipv6_global;
    }

    std::string Spiderip::get_spider_ipv6_global()
    {
        return spider_ipv6_global;
    }

    void Spiderip::set_spider_ipv6_unique_local(std::string spider_ipv6_unique_local)
    {
        this->spider_ipv6_unique_local = spider_ipv6_unique_local;
    }

    std::string Spiderip::get_spider_ipv6_unique_local()
    {
        return spider_ipv6_unique_local;
    }

    void Spiderip::set_spider_ipv6_link_local(std::string spider_ipv6_link_local)
    {
        this->spider_ipv6_link_local = spider_ipv6_link_local;
    }

    std::string Spiderip::get_spider_ipv6_link_local()
    {
        return spider_ipv6_link_local;
    }

    void Spiderip::set_spider_ipv6_link_local_scope_id(std::string spider_ipv6_link_local_scope_id)
    {
        this->spider_ipv6_link_local_scope_id = spider_ipv6_link_local_scope_id;
    }

    std::string Spiderip::get_spider_ipv6_link_local_scope_id()
    {
        return spider_ipv6_link_local_scope_id;
    }
}

File: Windows_csharp/SpiderIp.cs
/*
 * Title:  spider SpiderIp.cs (Windows csharp)
 * Author: Shuichiro Endo
 */

using System;
using System.Text;

namespace spider
{
    public class SpiderIp
    {
        private bool initFlag = false;
        private string spiderIpv4 = "";
        private string spiderIpv6Global = "";
        private string spiderIpv6UniqueLocal = "";
        private string spiderIpv6LinkLocal = "";
        private string spiderIpv6LinkLocalScopeId = "";

        public SpiderIp(string spiderIpv4,
                        string spiderIpv6Global,
                        string spiderIpv6UniqueLocal,
                        string spiderIpv6LinkLocal)
        {
            byte[] tmp;

            tmp = Encoding.UTF8.GetBytes(spiderIpv4.Trim());
            this.spiderIpv4 = Encoding.UTF8.GetString(tmp);

            tmp = Encoding.UTF8.GetBytes(spiderIpv6Global.Trim());
            this.spiderIpv6Global = Encoding.UTF8.GetString(tmp);

            tmp = Encoding.UTF8.GetBytes(spiderIpv6UniqueLocal.Trim());
            this.spiderIpv6UniqueLocal = Encoding.UTF8.GetString(tmp);

            tmp = Encoding.UTF8.GetBytes(spiderIpv6LinkLocal.Trim());
            this.spiderIpv6LinkLocal = Encoding.UTF8.GetString(tmp);

            if(!string.IsNullOrEmpty(spiderIpv6LinkLocal))
            {
                int percentIndex = spiderIpv6LinkLocal.IndexOf('%');
                if (percentIndex != -1)
                {
                    this.spiderIpv6LinkLocalScopeId = spiderIpv6LinkLocal.Substring(percentIndex + 1);
                    this.spiderIpv6LinkLocal = spiderIpv6LinkLocal.Substring(0, percentIndex);

                    initFlag = true;
                }else
                {
                    Console.WriteLine("[-] ipv6 link local address does not include a scope id");

                    initFlag = false;
                }

            }else
            {
                initFlag = true;
            }
        }

        ~SpiderIp()
        {

        }

        public bool InitFlag
        {
            get { return initFlag; }
            set { initFlag = value; }
        }

        public string SpiderIpv4
        {
            get { return spiderIpv4; }
            set { spiderIpv4 = value; }
        }

        public string SpiderIpv6Global
        {
            get { return spiderIpv6Global; }
            set { spiderIpv6Global = value; }
        }

        public string SpiderIpv6UniqueLocal
        {
            get { return spiderIpv6UniqueLocal; }
            set { spiderIpv6UniqueLocal = value; }
        }

        public string SpiderIpv6LinkLocal
        {
            get { return spiderIpv6LinkLocal; }
            set { spiderIpv6LinkLocal = value; }
        }

        public string SpiderIpv6LinkLocalScopeId
        {
            get { return spiderIpv6LinkLocalScopeId; }
            set { spiderIpv6LinkLocalScopeId = value; }
        }
    }
}

